src/components/RightSidebar/TableOfContents.tsx

```diff
@@ -19,28 +19,61 @@ const TableOfContents: FunctionalComponent<Props> = ({ headers = [], labels }) =
 		if (!toc.current) return;
 
 		const setCurrent: IntersectionObserverCallback = (entries) => {
+			const intersectingEntries = [];
+
 			for (const entry of entries) {
-				if (entry.isIntersecting) {
-					setCurrentID(entry.target.id);
-					break;
-				}
+				if (entry.isIntersecting) intersectingEntries.push(entry);
+			}
+
+			if (intersectingEntries.length === 1) {
+				setCurrentID(intersectingEntries[0].target.id);
+				scrollIntoView();
+			} else if (intersectingEntries.length > 1) {
+				const sortedEntries = intersectingEntries.sort((a, b) => {
+					return getHeadingIndex(a.target.id) - getHeadingIndex(b.target.id);
+				});
+				setCurrentID(sortedEntries[0].target.id);
+				scrollIntoView();
 			}
 		};
 
 		const observerOptions: IntersectionObserverInit = {
 			// Negative top margin accounts for `scroll-margin`.
 			// Negative bottom margin means heading needs to be towards top of viewport to trigger intersection.
-			rootMargin: '-100px 0% -66%',
+			rootMargin: '-100px 0px -40% 0px',
 			threshold: 1,
 		};
 
 		const headingsObserver = new IntersectionObserver(setCurrent, observerOptions);
 
+		// removes the headings from "on this page" on mobile using [id] selector.
+		const headings = document.querySelectorAll('article :is(h1,h2[id],h3)');
+
+		const getHeadingIndex = (id) => Array.from(headings).findIndex((heading) => heading.id === id);
+
+		let scrollPosition = 0;
+		let scrollDirection = 10;
+
+		const setScrollDirection = () => {
+			if (document.body.getBoundingClientRect().top > scrollPosition) scrollDirection = -10;
+			else scrollDirection = 10;
+			scrollPosition = document.body.getBoundingClientRect().top;
+		};
+
+		window.addEventListener('scroll', setScrollDirection);
+
+		const scrollIntoView = () => {
+			const scrollingContext = document.querySelector('.sidebar-nav-inner');
+			if (window.matchMedia('(prefers-reduced-motion)')) scrollingContext.scrollBy(0, scrollDirection);
+		};
```

Comment:
Wondering if we could simplify all this with the `scrollIntoView` browser API: https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView

Need to double check, but I bet we could use a `ref`, pass it to the current list item, and then scroll it each time it changes in another `useEffect`. Anything that avoids the `scroll` event seems like a good idea if possible.

(Guess we'd need to disable on mobile to avoid scrolling up to the table of contents.)

Reply:
I have tried to use `scrollIntoView`, it was my first attempt at implementing this, but it only scrolls the page scrollbar, not the sidebar one, so it doesn't help us in this case.

Reply:
Ah, you’re probably right! I think we could probably still move this logic to only run on intersection somehow rather than tracking scroll events/direction. Like: if a heading is current and it’s not in the viewport, scroll the parent container.